inginious/frontend/pages/course_admin/student_list.py

```diff
@@ -192,33 +198,48 @@ def post_audiences(self, course, data, active_tab, msg, error):
-                    audiences = json.loads(data["audiences"])
-
-                for index, new_audience in enumerate(audiences):
-                    # In case of file upload, no id specified
-                    new_audience['_id'] = new_audience['_id'] if '_id' in new_audience else 'None'
-
-                    # Update the audience
-                    audience, errors = self.update_audience(course, new_audience['_id'], new_audience)
-
-                    # If file upload was done, get the default audience id
-                    audienceid = audience['_id']
-                    errored_students += errors
-
-                if len(errored_students) > 0:
-                    msg["audiences"] = _("Changes couldn't be applied for following students :") + "<ul>"
-                    for student in errored_students:
-                        msg["audiences"] += "<li>" + student + "</li>"
-                    msg["audiences"] += "</ul>"
-                    error["audiences"] = True
-                elif not error:
-                    msg["audiences"] = _("Audience updated.")
+                    for user_id, field, role, description in csv_data:
+                        user_id = user_id.strip()
+                        field = field.strip()
+                        role = role.strip()
+                        description = description.strip()
+                        if field not in ["username", "email", "realname"] + list(auth_method_ids) \
+                                or role not in ["student", "tutors"]:
+                            continue
+                        if field != "username":
+                            user = self.database.users.find_one({field: user_id})
+                            user_id = user["username"] if user is not None else ""
+                            # TODO - Handle user not found
+                        query = {"courseid": course.get_id(), "description": description}
+
+                        existing_audience = self.database.audiences.find_one(query)
+                        if not existing_audience:
+                            insert = {"courseid": course.get_id(), "description": description,
+                                      "students": [user_id], "tutors": []} \
+                                if role == "student" else \
+                                {"courseid": course.get_id(), "description": description,
+                                 "students": [], "tutors": [user_id]}
+                            self.database.audiences.insert_one(insert)
+                        else:
+                            update = {'$push': {'students': user_id}} \
+                                if role == "student" else {'$push': {'tutors': user_id}}
+                            self.database.audiences.update_one(query, update)
+                        if user_id not in stud_list and role == "student":
+                            self.user_manager.course_register_user(course, user_id)
```

Comment:
Same remark, you can merge the student list with the list in database in one operation, here you will perform a query for each student.